src/librustc_metadata/decoder.rs

```diff
@@ -483,10 +484,15 @@ impl<'a, 'tcx> CrateMetadata {
     }
 
     pub fn get_def(&self, index: DefIndex) -> Option<Def> {
-        match self.is_proc_macro(index) {
-            true => Some(Def::Macro(self.local_def_id(index))),
-            false => self.entry(index).kind.to_def(self.local_def_id(index)),
+        if !self.is_proc_macro(index) {
+            return self.entry(index).kind.to_def(self.local_def_id(index));
+        }
+
+        if let Some(ref proc_macros) = self.proc_macros {
+            let kind = proc_macros[index.as_usize() - 1].1.kind();
+            return Some(Def::Macro(self.local_def_id(index), kind));
         }
+        unreachable!();
```

Comment:
nit: I think this would be clearer as an `if !self.is_proc_macro(index) { ... } else { ... }` to avoid early returns.

Also, I would do `let kind = self.proc_macros.as_ref().unwrap()[index.as_usize() - 1].1.kind();` in the `else` case to avoid `unreachable!()`.